### PR TITLE
TensorFlow Split and LRN operator support for Alexnet Model

### PR DESCRIPTION
**Pull request by dayanandasiet**
1. Tensorflow input parse logic updated
2. LRN bugfix
3. Split operator added
4. Input shapes parse logic updated
To support alexnet on tensorflow(https://www.cs.toronto.edu/~guerzhoy/tf_alexnet/)

Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from others in the community.

### CODE REVIEW

File: nnvm/python/nnvm/frontend/tensorflow.py [hunk 1]
```diff
@@ -420,18 +424,24 @@ def _impl(inputs, attr, params):
...
-        attr_new = {}
-        depth_radius = attr.get('depth_radius', 5)
-        size = (depth_radius * 2) + 1
-        attr_new['axis'] = 3 # Fix axis, NHWC format
-        attr_new['size'] = size
-        attr_new['bias'] = attr.get('bias', 1)
-        attr_new['alpha'] = attr.get('alpha', 1) * size
-        attr_new['beta'] = attr.get('beta', 0.5)
-        return AttrCvt(op_name='lrn')(new_inputs, attr_new)
+        depth_radius = attr.get('depth_radius', 2)
```
Comment (srkreddy1238) [inline nnvm/python/nnvm/frontend/tensorflow.py]: LRN is already implemented earlier. Any reason for this modify ?

File: nnvm/python/nnvm/frontend/tensorflow.py [hunk 2]
```diff
@@ -142,11 +143,11 @@ def _impl(inputs, attr, params):
             stride_h, stride_w = attr['strides']
             kernel_h, kernel_w = attr['kernel_shape']
             if attr['data_format'] == 'NHWC':
-                in_h = input_shapes[0][1]
-                in_w = input_shapes[0][2]
+                in_h = input_shapes[1]
```
Comment (srkreddy1238) [inline nnvm/python/nnvm/frontend/tensorflow.py]: input_shapes -> input_shape

File: nnvm/python/nnvm/frontend/tensorflow.py [hunk 3]
```diff
@@ -174,25 +175,28 @@ def _impl(inputs, attr, params):
         input_shapes = attr['_input_shapes'][inputs[0]]
 
         # Extract kernel shape from params
-        conv_param_weights = params[inputs[1].list_output_names()[0]]
+        if inputs[1] in attr['_input_shapes']:
```
Comment (srkreddy1238) [inline nnvm/python/nnvm/frontend/tensorflow.py]: Can you share details of use case where convolution weights not in params ?

File: nnvm/python/nnvm/frontend/tensorflow.py [hunk 4]
```diff
@@ -420,18 +424,24 @@ def _impl(inputs, attr, params):
             ignores=['index_type', 'T'])(new_inputs, attr)
     return _impl
 
+def _split():
+    def _impl(inputs, attr, params):
+        pop_node = inputs.pop(0)
+        axis = params[pop_node.list_output_names()[0]].asnumpy()[0]
+        return AttrCvt(op_name="split",
+                       ignores=['num_split'],
```
Comment (srkreddy1238) [inline nnvm/python/nnvm/frontend/tensorflow.py]: Suggest to use transforms instead of ignores & extras.

File: nnvm/python/nnvm/frontend/tensorflow.py [hunk 5]
```diff
@@ -974,24 +985,28 @@ def from_tensorflow(self, graph):
...
+                    if len(node_input_key) > 1:
+                        slot_num = int(node_input_key[1])
+                    node_input_key = node_input_key[0]
+
+                    try:
+                        try:
+                            input_sym = self._nodes[node_input_key].__getitem__(slot_num)
+                        except NNVMError:
+                            # TODO: Fancy node name invalid slot neglect
+                            input_sym = self._nodes[node_input_key].__getitem__(0)
```
Comment (srkreddy1238) [inline nnvm/python/nnvm/frontend/tensorflow.py]: Can you share details on this fall back?
I see slot_num is initialized to 0 above already.

File: nnvm/python/nnvm/frontend/tensorflow.py [hunk 6]
```diff
@@ -974,24 +985,28 @@ def from_tensorflow(self, graph):
...
+
+                        inputs.append(input_sym)
+                        input_shapes[input_sym] = self._output_shapes[
+                            node_input_key].__getitem__(slot_num)
+                        attr['_input_shapes'] = input_shapes
+                    except KeyError:
```
Comment (srkreddy1238) [inline nnvm/python/nnvm/frontend/tensorflow.py]: This except is for the same statement below where the ```self._nodes``` doesn't have key ```node_input_key```
```input_sym = self._nodes[node_input_key].__getitem__(slot_num)```

Suggest to make one try block with multiple except instead of nested.

File: nnvm/tests/python/frontend/tensorflow/test_forward.py [hunk 7]
```diff
@@ -925,39 +925,69 @@ def _get_sample(data, state):
...
+    np_data = np.random.uniform(size=ip_shape).astype(dtype)
+    with tf.Session() as sess:
+        final_graph_def = tf.graph_util.convert_variables_to_constants(
+            sess,
+            sess.graph.as_graph_def(add_shapes=True),
+            ['split'])
+        tf_output = run_tf_graph(sess, [np_data], ['in_data:0'], 'split:0')
+        tvm_output = run_tvm_graph(final_graph_def, [np_data], ['in_data'],
+                                   tf_output.shape, dtype)
+        np.testing.assert_allclose(tf_output, tvm_output, atol=1e-5, rtol=1e-5)
```
Comment (srkreddy1238) [inline nnvm/tests/python/frontend/tensorflow/test_forward.py]: Split results in multiple outputs, hence  suggest to enhance run methods to return multiple outputs and compare the same.

File: nnvm/tests/python/frontend/tensorflow/test_forward.py [hunk 8]
```diff
@@ -925,39 +925,69 @@ def _get_sample(data, state):
...
+        sess.close()
 
-def _test_lrn(ishape, size, axis, bias, alpha, beta):
-    """ testing local response normalization """
-    lrn_depth_radius = size / 2
+def test_forward_split():
+    '''test split operator'''
+    _test_split((2, 3), 2, axis=0)
+    _test_split((6, 3), 3, axis=0)
```
Comment (srkreddy1238) [inline nnvm/tests/python/frontend/tensorflow/test_forward.py]: Pls add a case for -ve axis

File: nnvm/python/nnvm/frontend/tensorflow.py [hunk 9]
```diff
@@ -420,18 +424,24 @@ def _impl(inputs, attr, params):
             ignores=['index_type', 'T'])(new_inputs, attr)
     return _impl
 
+def _split():
+    def _impl(inputs, attr, params):
+        pop_node = inputs.pop(0)
```
Comment (srkreddy1238) [inline nnvm/python/nnvm/frontend/tensorflow.py]: Ref. [Split](https://www.tensorflow.org/api_docs/python/tf/split)
tensorflow has two variations for split as ```num_split``` and ```size_split```.
Please handle if possible or raise exception accordingly.

If handling size_split please handle ```num``` attribute if possible or raise exception accordingly.

File: nnvm/python/nnvm/frontend/tensorflow.py [hunk 10]
```diff
@@ -174,25 +175,28 @@ def _impl(inputs, attr, params):
         input_shapes = attr['_input_shapes'][inputs[0]]
 
         # Extract kernel shape from params
-        conv_param_weights = params[inputs[1].list_output_names()[0]]
+        if inputs[1] in attr['_input_shapes']:
+            conv_param_weights = tuple(attr['_input_shapes'][inputs[1]])
+        else:
+            conv_param_weights = params[inputs[1].list_output_names()[0]].shape
```
Comment (srkreddy1238) [inline nnvm/python/nnvm/frontend/tensorflow.py]: conv_param_weights -> weights_shape (it's no more a param hence.)

File: nnvm/python/nnvm/frontend/tensorflow.py (hunk 1 above)
Comment (dayanandasiet) [reply nnvm/python/nnvm/frontend/tensorflow.py]: Previously implemented LRN was wrong as per TF formula and input tensor is never pass to operator, this leak observe after run the [AlexNet](https://www.cs.toronto.edu/~guerzhoy/tf_alexnet/) model

File: nnvm/python/nnvm/frontend/tensorflow.py (hunk 3 above)
Comment (dayanandasiet) [reply nnvm/python/nnvm/frontend/tensorflow.py]: This case is observe while running [AlexNet](https://www.cs.toronto.edu/~guerzhoy/tf_alexnet/) model.

Comment (srkreddy1238) [reply nnvm/python/nnvm/frontend/tensorflow.py]: Is weights output of another node in runtime?

Can u share serialised tensor flow graph?

File: nnvm/python/nnvm/frontend/tensorflow.py (hunk 5 above)
Comment (dayanandasiet) [reply nnvm/python/nnvm/frontend/tensorflow.py]: Some of Layer implementation in frontend(LSTM, Ptb) this assign fancy node name (eg. Model/RNN/cell_0/RnnCell:6 ), in this case 6 is slot updated as per our logic and only one node will be present, in this fancy naming convention we need to parse the default(0) node.

File: nnvm/python/nnvm/frontend/tensorflow.py (hunk 6 above)
Comment (dayanandasiet) [reply nnvm/python/nnvm/frontend/tensorflow.py]: as address the [review](https://github.com/dmlc/tvm/pull/1572#discussion_r208894072) comment, we need nested try block to handle fancy node name convention logic.

File: nnvm/python/nnvm/frontend/tensorflow.py (hunk 1 above)
Comment (srkreddy1238) [reply nnvm/python/nnvm/frontend/tensorflow.py]: I think that was a bug passing empty inputs ```new_inputs = []```
Removing new_inputs and using inputs instead should work for alexnet as well.
Keep the remaining implementation default value handling.

@siju-samuel  can you confirm the depth_radius default value as 5 which results size=11 ? or is it a typo !!

```
   depth_radius = attr.get('depth_radius', 5)
   size = (depth_radius * 2) + 1
``` 

Test case was passed as the input was taken from params instead.
#1546 should fix this issue by replacing all const inputs to Placeholders.

File: nnvm/python/nnvm/frontend/tensorflow.py (hunk 6 above)
Comment (srkreddy1238) [reply nnvm/python/nnvm/frontend/tensorflow.py]: It's the same statement which cause 2 exceptions in two different cases, one try on ```input_sym = self._nodes[node_input_key].__getitem__(slot_num)``` with two ```except``` should work. Pls check.

File: nnvm/python/nnvm/frontend/tensorflow.py (hunk 1 above)
Comment (siju-samuel) [reply nnvm/python/nnvm/frontend/tensorflow.py]: @srkreddy1238  as per tensorflow, the default value of depth_radius is 5. So the size will become 11.
https://www.tensorflow.org/api_docs/python/tf/nn/local_response_normalization

File: nnvm/python/nnvm/frontend/tensorflow.py (hunk 9 above)
Comment (dayanandasiet) [reply nnvm/python/nnvm/frontend/tensorflow.py]: [Split](https://www.tensorflow.org/versions/r1.0/api_docs/cc/class/tensorflow/ops/split) had num_split attribute and [SplitV](https://www.tensorflow.org/versions/r1.0/api_docs/cc/class/tensorflow/ops/split-v) had size_splits attribute. As of this PR supported [Split](https://www.tensorflow.org/versions/r1.0/api_docs/cc/class/tensorflow/ops/split-v) operator

File: nnvm/python/nnvm/frontend/tensorflow.py (hunk 3 above)
Comment (dayanandasiet) [reply nnvm/python/nnvm/frontend/tensorflow.py]: fine tuned [Alexnet](https://kratzert.github.io/2017/02/24/finetuning-alexnet-with-tensorflow.html) can be found here for the above reference

File: nnvm/tests/python/frontend/tensorflow/test_forward.py (hunk 8 above)
Comment (dayanandasiet) [reply nnvm/tests/python/frontend/tensorflow/test_forward.py]: [Split](https://www.tensorflow.org/versions/r1.0/api_docs/cc/class/tensorflow/ops/split) axis should be range (0, rank(value)). so negative axis is not added

File: nnvm/tests/python/frontend/tensorflow/test_forward.py (hunk 7 above)
Comment (dayanandasiet) [reply nnvm/tests/python/frontend/tensorflow/test_forward.py]: [Split](https://www.tensorflow.org/versions/r1.0/api_docs/cc/class/tensorflow/ops/split) returns the split result with list so comparing with List content result.

File: nnvm/python/nnvm/frontend/tensorflow.py (hunk 6 above)
Comment (dayanandasiet) [reply nnvm/python/nnvm/frontend/tensorflow.py]: First try block we are trying to access the slot base input and after caught NNVMError we are trying to access default(0) location input, in this case if we caught error should be thrown to user.

File: nnvm/tests/python/frontend/tensorflow/test_forward.py (hunk 7 above)
Comment (srkreddy1238) [reply nnvm/tests/python/frontend/tensorflow/test_forward.py]: run_tf_graph is not designed to result a list. Also we pass 'split:0' to return only first result.
run_tvm_graph can return list if output_shape, output_dtypes are lists.

Please check.

Comment (dayanandasiet) [reply nnvm/tests/python/frontend/tensorflow/test_forward.py]: Yeah missed scenario, will work on it.

**Comment by dayanandasiet** [comment]
thanks @srkreddy1238.

Reworked for accepted review comments. Welcome more review from @yzhliu

File: nnvm/tests/python/frontend/tensorflow/test_forward.py [hunk 11]
```diff
@@ -71,16 +71,23 @@ def run_tvm_graph(graph_def, input_data, input_node, output_shape, output_dtype)
...
         input_dict = {}
         for i, e in enumerate(input_node):
             input_dict[e] = input_data[i]
     else:
         input_dict = {input_node: input_data}
 
-    output_data = sess.run(tensor, input_dict)
+    if isinstance(output_node, list):
+        output_list = []
+        for node_name in output_node:
```
Comment (srkreddy1238) [inline nnvm/tests/python/frontend/tensorflow/test_forward.py]: Can run one time by passing list of tensors to run method.

Comment (dayanandasiet) [reply nnvm/tests/python/frontend/tensorflow/test_forward.py]: Address the problem with update rework solution.

File: nnvm/python/nnvm/frontend/tensorflow.py (hunk 4 above)
Comment (yzhliu) [reply nnvm/python/nnvm/frontend/tensorflow.py]: Have you reached consensus the changes here is good?

**Comment by yzhliu** [comment]
@dayanandasiet  Can you confirm the attr implement (where @srkreddy1238 comment 'Suggest to use transforms instead of ignores & extras') is good? Otherwise LGTM. Thanks.

File: nnvm/tests/python/frontend/tensorflow/test_forward.py [hunk 12]
```diff
@@ -787,29 +789,72 @@ def _get_sample(data, state):
...
-                                            depth_radius=lrn_depth_radius,
-                                            bias=bias,
-                                            alpha=alpha,
-                                            beta=beta)
 
-        compare_tf_with_tvm(inp_array, 'lrn0_data:0', 'lrn:0')
+#######################################################################
+# LRN (Local Response Normalization)
+# ----------------------------------
+def _test_lrn(ip_shape, depth_radius=2, alpha=1e-05, beta=0.75, bias=1.0):
```
Comment (srkreddy1238) [inline nnvm/tests/python/frontend/tensorflow/test_forward.py]: [Ref.](https://www.tensorflow.org/api_docs/python/tf/nn/local_response_normalization)  we don't need to explicitly set the default values. TF can handle them if not set.

Suggest to ignore to allow TF default values play role here.

**Comment by yuruofeifei** [comment]
Hi @dayanandasiet , 
Do you have any update on this PR?

**Comment by yzhliu** [comment]
Any update @dayanandasiet ?

**Comment by yuruofeifei** [comment]
Hi @dayanandasiet , do you have any update on this PR?
I'm currently working on TF convertor which will need split operator.

**Comment by dayanandasiet** [comment]
Will fix review comment and push this changes with new PR. As of now closing this PR.
